pilot/pkg/xds/endpoints/endpoint_builder.go

```diff
@@ -788,3 +822,59 @@ func getSubSetLabels(dr *v1alpha3.DestinationRule, subsetName string) labels.Ins
 
 	return nil
 }
+
+// For services that have a waypoint, we want to send to the waypoints rather than the service endpoints.
+// Lookup the
+func (b *EndpointBuilder) findServiceWaypoint(endpointIndex *model.EndpointIndex) ([]*model.IstioEndpoint, bool) {
+	if b.nodeType != model.Router {
+		// Currently only ingress will call waypoints
+		return nil, false
+	}
+	// ...and they must explicitly opt in
+	if b.service.Attributes.Labels["istio.io/ingress-use-waypoint"] != "true" {
+		return nil, false
+	}
+	if b.service.GetAddressForProxy(b.proxy) == constants.UnspecifiedIP {
+		// No VIP, so skip this. Currently, waypoints can only accept VIP traffic
+		return nil, false
+	}
+
+	svcs := b.push.ServicesWithWaypoint(b.service.Attributes.Namespace + "/" + string(b.hostname))
+	if len(svcs) == 0 {
+		// Service isn't captured by a waypoint
+		return nil, false
+	}
+	if len(svcs) > 1 {
```

Comment:
Trying to think how this will interact with wildcard hostnames for services. What will be the expectation if, for example, there are 2 wildcard services (one more specific than the other) and they're both served by the same waypoint? Do we need to care about matching rules at this level?

Reply:
The hostname is a key on the cluster regardless of if there wildcards? Like `*.com` is a distinct cluster mapped to a distinct service for `*.com`, its not a wildcard match but an exact match. IIUC the question right

Reply:
Ok that's what I thought but just checking. After dealing with the VS logic, wildcards make me nervous lol